### PR TITLE
Integration test for out of order tx execution between proposer and validator

### PR DESCRIPTION
**Pull request by iulianpascalau**
~added a long integration test regarding sending move balance transactions from an account that receives rewards~
added integration test to test edge cases of out of order processing between proposer and validator

### CODE REVIEW

File: integrationTests/longTests/txsFromGatheredRewards_test.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+	"github.com/ElrondNetwork/elrond-go/sharding"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+var log = logger.GetOrCreate("integrationtests/singleshard/block")
+
+// TestShardShouldNotProposeAndExecuteTwoBlocksInSameRound tests what happens in a setting where a proposer/validator
+// periodically sends transactions to someone else. The tests starts with the node having 0 balance.
+func TestExecutingTransactionsFromGatheredRewards(t *testing.T) {
```
Comment (sasurobert) [inline integrationTests/longTests/txsFromGatheredRewards_test.go]: function too long. over 100 lines.

File: process/transaction/shardProcess.go [hunk 2]
```diff
@@ -112,6 +113,11 @@ func (txProc *txProcessor) ProcessTransaction(tx *transaction.Transaction) error
 		return err
 	}
 
+	log := logger.GetOrCreate("aaa")
+	txHash, _ := core.CalculateHash(txProc.marshalizer, txProc.hasher, tx)
```
Comment (sasurobert) [inline process/transaction/shardProcess.go]: delete these lines.

Comment (iulianpascalau) [reply process/transaction/shardProcess.go]: deleted

File: integrationTests/longTests/txsFromGatheredRewards_test.go (hunk 1 above)
Comment (iulianpascalau) [reply integrationTests/longTests/txsFromGatheredRewards_test.go]: will refactor

Comment (iulianpascalau) [reply integrationTests/longTests/txsFromGatheredRewards_test.go]: done

File: integrationTests/longTests/txsFromGatheredRewards_test.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+		}
+	}
+}
+
+func checkSameBlockHeight(t *testing.T, nodesMap map[uint32][]*integrationTests.TestProcessorNode) {
+	for _, nodes := range nodesMap {
+		crtBlock := nodes[0].BlockChain.GetCurrentBlockHeader()
+		for _, n := range nodes {
+			blkc := n.BlockChain.GetCurrentBlockHeader()
+			require.False(t, (crtBlock == nil) != (blkc == nil))
```
Comment (bogdan-rosianu) [inline integrationTests/longTests/txsFromGatheredRewards_test.go]: please add an explanation to this `require`. Also you can move line 116 before the `for` and maybe rename:
`crtBlock` -> `referenceBlock`
`blkc` -> `crtBlock`

File: integrationTests/longTests/txsFromGatheredRewards_test.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+			require.Equal(t, crtBlock.GetNonce(), blkc.GetNonce())
+		}
+	}
+}
+
+func printBalance(node *integrationTests.TestProcessorNode) {
```
Comment (bogdan-rosianu) [inline integrationTests/longTests/txsFromGatheredRewards_test.go]: you don't print just the balance so please rename this function

File: integrationTests/longTests/txsFromGatheredRewards_test.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+		"nonce", accnt.GetNonce(),
+		"balance", accnt.(*state.Account).Balance,
+	)
+}
+
+func getBlockProposersIndexes(
+	consensusMap map[uint32][]*integrationTests.TestProcessorNode,
+	nodesMap map[uint32][]*integrationTests.TestProcessorNode,
+) map[uint32]int {
+
```
Comment (bogdan-rosianu) [inline integrationTests/longTests/txsFromGatheredRewards_test.go]: remove empty line. it is a linter issue

File: integrationTests/longTests/txsFromGatheredRewards_test.go [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+func generateAndSendTxs(
+	n *integrationTests.TestProcessorNode,
+	transferValue uint64,
+	sk crypto.PrivateKey,
+	addr state.AddressContainer,
+	pkReceiver crypto.PublicKey,
+) {
+
```
Comment (bogdan-rosianu) [inline integrationTests/longTests/txsFromGatheredRewards_test.go]: remove empty line

File: integrationTests/longTests/txsFromGatheredRewards_test.go (hunk 1 above)
Comment (iulianpascalau) [reply integrationTests/longTests/txsFromGatheredRewards_test.go]: comment updated

File: integrationTests/longTests/txsFromGatheredRewards_test.go [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+				firstNode,
+				transferValue,
+				firstNode.OwnAccount.SkTxSign,
+				firstNode.OwnAccount.Address,
+				nodesMap[sharding.MetachainShardId][0].OwnAccount.PkTxSign)
+			time.Sleep(time.Second)
+		}
+	}()
+
+	nbBlocksProduced := uint64(20)
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply integrationTests/longTests/txsFromGatheredRewards_test.go]: done

File: integrationTests/longTests/txsFromGatheredRewards_test.go (hunk 3 above)
Comment (iulianpascalau) [reply integrationTests/longTests/txsFromGatheredRewards_test.go]: done

File: integrationTests/longTests/txsFromGatheredRewards_test.go (hunk 4 above)
Comment (iulianpascalau) [reply integrationTests/longTests/txsFromGatheredRewards_test.go]: changed to `printAccount`

File: integrationTests/longTests/txsFromGatheredRewards_test.go (hunk 5 above)
Comment (iulianpascalau) [reply integrationTests/longTests/txsFromGatheredRewards_test.go]: done

File: integrationTests/longTests/txsFromGatheredRewards_test.go (hunk 6 above)
Comment (iulianpascalau) [reply integrationTests/longTests/txsFromGatheredRewards_test.go]: done

File: integrationTests/multiShard/block/edgecases_test.go [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+
+	firstNode := nodesMap[0][0]
+	secondNode := nodesMap[0][1]
+	receiver := nodesMap[sharding.MetachainShardId][0].OwnAccount.PkTxSign
+
+	transferValue := integrationTests.MinTxGasLimit * integrationTests.MinTxGasPrice
+
+	go func() {
+		for {
+			generateAndSendTxs(
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply integrationTests/multiShard/block/edgecases_test.go]: done
